chore: [MR-640] Make request metadata non optional

stiegerc opened the pull request:
With old certification versions removed (i.e. we will never go back to a version without metadata) and no old requests without metadata are still around (checked via metrics), it seems reasonable to ditch `Option` here.

There are no very old requests in
- output queues because those time out after 5 minutes.
- in streams because we know no stream stalled for that long (there are alerts about this)
- in input queues because the metrics show that the number of messages went to 0 at least once on all subnets since the metadata was introduced.

stiegerc commented:
We could also remove the [optional](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/protobuf/def/state/queues/v1/queues.proto?L57) in the protobuf definition as well, but I don't think it matters that much there.

I would however leave the [optional](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/canonical_state/src/encoding/types.rs?L127) for the canonical state since changing that is quite involved with a new certification version and so on and IIUC it won't affect encoding it to CBOR since it's just skipped in the case of `None`.

stiegerc commented on rs/types/types/src/messages.rs:
```diff
@@ -763,27 +763,19 @@ mod tests {
     #[test]
     fn serialize_request_via_bincode() {
-        for metadata in [
-            None,
-            Some(RequestMetadata::new(
```
The case of `None` here caused a hard to track bug which is why this test was introduced. The version for responses below was just for completeness. I'd say it doesn't hurt to keep it around though.

alin-at-dfinity commented on rs/types/types/src/messages/inter_canister.rs:
```diff
@@ -633,9 +641,9 @@ impl From<pb_queues::RequestMetadata> for RequestMetadata {
-            call_tree_start_time: Time::from_nanos_since_unix_epoch(
-                metadata.call_tree_start_time_nanos.unwrap_or(0),
-            ),
+            call_tree_start_time: metadata
+                .call_tree_start_time_nanos
+                .map_or(UNIX_EPOCH, Time::from_nanos_since_unix_epoch),
```
I'd add a TODO here (backed by a ticket) to replace this with `try_from_option_field()` in a future release. There should be no more `None` values in the checkpoint after this.

stiegerc replied:
I deprecated the optional fields and added non-optional ones. Added a todo to remove the optional fields with a [ticket](https://dfinity.atlassian.net/jira/software/c/projects/MR/boards/223?selectedIssue=MR-641).

alin-at-dfinity reviewed (approved):
> We could also remove the [optional](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/protobuf/def/state/queues/v1/queues.proto?L57) in the protobuf definition as well, but I don't think it matters that much there.

The difference between having and not having those two `optional` qualifiers there is whether zero values are encoded as zero (with `optional`) or not encoded at all (without the `optional`). So unless the presubmit check complains, I would drop it. We do not care about the difference between `Some(0)` and `None` (we will interpret it as `0` in either case).

> I would however leave the [optional](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/canonical_state/src/encoding/types.rs?L127) for the canonical state since changing that is quite involved with a new certification version and so on and IIUC it won't affect encoding it to CBOR since it's just skipped in the case of None.

I would create a follow-up ticket. Not necessarily do something about it immediately, but just as a way to remember that this can be done at a future time.

stiegerc commented on rs/system_api/src/sandbox_safe_system_state.rs:
```diff
@@ -353,7 +353,7 @@ impl SystemStateChanges {
                 .requests
                 .first()
-                .and_then(|request| request.metadata.clone()),
+                .map_or_else(Default::default, |request| request.metadata.clone()),
```
The Rust [documentation](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) says `map_or_else()` is lazily evaluated so it shouldn't matter.

stiegerc replied:
If you don't mind I'll leave it as it is.

alin-at-dfinity replied:
Right, I missed that the first argument was `Default::default` (as opposed to `Default::default()`).

Apologies for the noise.

stiegerc commented on rs/canonical_state/src/encoding/tests/compatibility.rs:
```diff
@@ -591,13 +585,13 @@ fn canonical_encoding_request_with_u128_cycles() {
 
         assert_eq!(
-            "A1 00 A6 00 4A 00 00 00 00 00 00 00 01 01 01 01 4A 00 00 00 00 00 00 00 02 01 01 02 03 03 A1 00 A2 00 1B C3 73 E0 EE 4E 3F 0A D2 01 1B 00 00 00 01 8E E9 0F F6 04 64 74 65 73 74 05 41 06",
+            "A1 00 A7 00 4A 00 00 00 00 00 00 00 01 01 01 01 4A 00 00 00 00 00 00 00 02 01 01 02 03 03 A1 00 A2 00 1B C3 73 E0 EE 4E 3F 0A D2 01 1B 00 00 00 01 8E E9 0F F6 04 64 74 65 73 74 05 41 06 07 A2 00 00 01 00",
```
Done.

alin-at-dfinity commented on rs/canonical_state/src/encoding/tests/compatibility.rs:
```diff
@@ -573,6 +573,8 @@ fn canonical_encoding_request_v18_plus() {
 ///       05                      # field_index(Request::method_payload)
 ///       41                      # bytes(1)
 ///          06                   # "\x06"
+///       07                        # unsigned(7)
+///       A2                        # map(2)
```
I believe this should be (but you should double-check):

```suggestion
///       07                      # field_index(Request::metadata)
///       A0                      # map(0)
```

stiegerc replied:
Done.

stiegerc commented:
> > We could also remove the [optional](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/protobuf/def/state/queues/v1/queues.proto?L57) in the protobuf definition as well, but I don't think it matters that much there.
> 
> The difference between having and not having those two `optional` qualifiers there is whether zero values are encoded as zero (with `optional`) or not encoded at all (without the `optional`). So unless the presubmit check complains, I would drop it. We do not care about the difference between `Some(0)` and `None` (we will interpret it as `0` in either case).
> 
> > I would however leave the [optional](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/canonical_state/src/encoding/types.rs?L127) for the canonical state since changing that is quite involved with a new certification version and so on and IIUC it won't affect encoding it to CBOR since it's just skipped in the case of None.
> 
> I would create a follow-up ticket. Not necessarily do something about it immediately, but just as a way to remember that this can be done at a future time.

Removed the optional fields from the protobuf versions. Added a todo with a [ticket](https://dfinity.atlassian.net/browse/MR-642) for the canonical version.

I also changed the _skip serializing if_ thing to skip it if it is `None` or `Some(0)`. While this won't matter for the call tree start time since that is never zero unless initialized with the default, but since the call tree depth is zero for initial calls (i.e. not downstream calls) and since these make up the vast bulk of messages, this seems quite significant. This is also how it would be without the option anyway.

alin-at-dfinity commented on rs/canonical_state/src/encoding/types.rs:
```diff
@@ -100,11 +100,12 @@ pub struct RequestOrResponse {
-    #[serde(skip_serializing_if = "Option::is_none")]
+    // TODO(MR-642): Remove `Option` from `call_tree_depth` and `call_tree_start_time`.
+    #[serde(skip_serializing_if = "is_none_or_zero")]
     pub call_tree_depth: Option<u64>,
-    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(skip_serializing_if = "is_none_or_zero")]
     pub call_tree_start_time_u64: Option<u64>,
-    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(skip_serializing_if = "is_none_or_zero")]
     pub call_subtree_deadline_u64: Option<u64>,
```
I'm not sure that replacing `Option::is_none` with `is_none_or_zero` is safe. We are changing the encoding of a `Request` without changing the certification version.

Both `canonical_encoding_request_with_u128_cycles()` and the changing hashes in `test_backward_compatibility()` are warning us against this.

stiegerc replied:
The hash in `canonical_encoding_request_with_u128_cycles()` still changes since we are hashing something else. I don't think it's a problem but one thing I could do would be to check if the metadata is the default on the coversion to the canonical version and then use `None` instead of Some(Default::default())`. I didn't do that because that will never be the case anyway except for the odd very old call perhaps.

alin-at-dfinity commented on rs/protobuf/def/state/queues/v1/queues.proto:
```diff
@@ -54,8 +54,11 @@ message StreamEntry {
 message RequestMetadata {
-  optional uint64 call_tree_depth = 1;
-  optional uint64 call_tree_start_time_nanos = 2;
+  // TODO(MR-641): Remove deprecated fields.
+  optional uint64 deprecated_opt_call_tree_depth = 1;
+  optional uint64 deprecated_opt_call_tree_start_time_nanos = 2;
+  uint64 call_tree_depth = 4;
+  uint64 call_tree_start_time_nanos = 5;
```
I don't think this is necessary. The difference between `optional uint64` and `uint64` is that the former explicitly encodes a value of zero, whereas the latter just skips encoding it (producing zero bytes).

So from the point of view of the encoded data, it should be entirely safe to just drop the `optional` qualifier.

stiegerc replied:
I get one failing test that may be due to me trying it with just removing `optional` here: //pre-commit:buf-breaking
It passes when I run it locally though. Not sure what to make of it.

stiegerc replied:
The exception has fixed the tests. I take it this means it should be ok to just remove the 'optional' from the protobuf definitions.

alin-at-dfinity commented on rs/types/types/src/messages/inter_canister.rs:
```diff
@@ -612,10 +612,11 @@ impl From<Response> for RequestOrResponse {
-                metadata.call_tree_start_time.as_nanos_since_unix_epoch(),
-            ),
+            // TODO(MR-641): Remove deprecated fields.
+            deprecated_opt_call_tree_depth: None,
+            deprecated_opt_call_tree_start_time_nanos: None,
```
Also, this is not backwards-compatible. If we rolled back a replica running this code, the old replica version would decode all `RequestMetadatas` encoded this way as `(0, 0)`.

eichhorl commented on buf.yaml:
```diff
@@ -21,6 +21,7 @@ lint:
     - DEFAULT
 breaking:
   ignore:
+    - state/queues/v1/queues.proto
```
I think instead of adding this ignore, you could also add the `CI_OVERRIDE_BUF_BREAKING` label to this PR to skip the buf check.

stiegerc replied:
Ack. But I'm removing this right after this is merged anyway.
